### ISSUE TITLE
Infinite Scrollback

### ISSUE THREAD

**Issue by iancward** [microsoft/terminal#1410]
# Summary of the new feature/enhancement
I know at least gnome-terminal supports infinite scrollback as a checkbox option in profile settings; it would be great to have that here (I'd love to be able to replace X server+gnome-terminal for WSL).

I checked to make sure that there wasn't an issue for this already, and I saw that you have a TODO comment in the code for this, so I figured I'd open a GitHub issue so that you can use the 👍 to figure out how to rank it.

# Proposed technical implementation details (optional)
Looks like the plan is for users to set HistorySize to `-1`
https://github.com/microsoft/terminal/blob/900d0c3cce39fa191e16b0224fa32c3441f8de24/src/cascadia/TerminalCore/Terminal.cpp#L88

**Comment by JushBJJ** [comment]
Would the buffer content be stored inside a file?

**Comment by egmontkob** [comment]
In GNOME Terminal (VTE) we store the scrollback in a compressed and encrypted file. Storing it in memory could easily result in heavy swapping, followed by the OOM killer kicking in. We store it on disk even if it's configured to have a finite size.

The file is unlinked immediately after creating it, so it's freed up in the unlikely case that gnome-terminal (or another VTE-based terminal) crashes. During normal operation, the file handle is released (and thus the file contents get freed up) and encryption key is explicitly zeroed out in the memory when the terminal tab/window is closed.

We used to store it in unencrypted file, but people were worried about privacy / data leakage issues e.g. if the computer gets stolen, see e.g. a [generic security report](http://www.climagic.org/bugreports/libvte-scrollback-written-to-disk.html) and the [tracking bug](https://bugzilla.gnome.org/show_bug.cgi?id=664611). (A nice extension would be if we could detect if the file system is encrypted and then skip this step.)

Konsole, on the other hand, warns you in its config dialog that "When using [unlimited scrollback], the scrollback data will be written unencrypted to temporary files." If fixed scrollback is configured, Konsole stores it in memory (and again, informs the user about it and the corresponding risks).

**Comment by davidbIQ** [comment]
As an easy first step is to allow a rightclick on the tab and have an option there for scrollback lines that would just apply to that tab.  Then it's all in memory and fully up to the user to choose a reasonable amount (some in memory compression might be nice but likely optional).  It gets around all the security issues, and other issues if you default it to every terminal that are left open for weeks - you could auto-change the tab color so they're aware that it's different than normal.

**Comment by naikel** [comment]
This is a **massive** change.

The TextBuffer is limited to COORD coordinates which they are of short type (32767 max).  Only to enable "virtual" coordinates of the TextBuffer means to change more than half of the code.

So I guess the first approach would be to change COORD to til::point, SIZE to til::size, and RECT & SMALL_RECT to... well good question.  SMALL_RECT is inclusive and that could be translated to a til::rectangle.  But RECT?  Then you would have to call Viewport::ToInclusive() ?

It's easier to first support TextBuffer >= 32768 lines and then implement the infinite scrollback.

Buy from my perspective I think everything has to be changed, GDI renderers, UI Automation types, everything.

**Comment by j4james** [comment]
I don't think it should be necessary to replace the short ints everywhere (you're not going to be able to change the parameters of the legacy console APIs anyway). I reckon you could leave the in-memory buffer with a short limit, and just persist the oldest lines to disk whenever it is about to cycle. Then the only major complication is getting the scrollback viewer to page that data back into memory when the user scrolls past the top of the buffer. That's got to be easier than trying to rewrite _everything_ with new types.

**Comment by naikel** [comment]
>  I reckon you could leave the in-memory buffer with a short limit, and just persist the oldest lines to disk whenever it is about to cycle.

You still need the virtual position to be at least a LONG (but a til::point will be better).  The Viewport "moves" around this TextBuffer using these coordinates.  Even if the buffer is not complete in memory and it is on disk, you need to know where in the buffer you are. The resulting screen has to be SHORT since like you're saying it is calling an API that only accepts that.

That still means changing everything except probably GdiEngine.

**Comment by j4james** [comment]
> You still need the virtual position to be at least a LONG

But that's the only thing that needs to be LONG. The in-memory buffer really just needs to be the size of the visible viewport, with the relevant section of the on-disk buffer loaded into it.

Anyway, this was just a suggestion. Whoever is actually implementing this feature will no doubt have their own ideas on how to approach the problem. I just would recommend trying rewrite the entire conhost framework unless it was absolutely necessary, and I don't think it is.

**Comment by DSloaneNGPVAN** [comment]
Infinity is a really big number - personally I would love to see a 10x or 100x increase, from the current 32767 max!
And/or a reference to 32767 in the https://docs.microsoft.com/en-us/windows/terminal/customize-settings/profile-advanced#history-size documentation. 
And/or an max value validation in the historySize field under Advanced Settings.

**Comment by davidbIQ** [comment]
yea, the way putty does it works very well, its default of 2000 lines and then if you need more you can set it higher I think I've put 2M in there before I know with certainty 200,000 works fine which is plenty.  Regarding security of disk writes, I think if the windows not open the file not existing would be fine but I'm not a security expert, if it's an issue having it be encrypted shouldn't be much of an additional step as only that window would need to understand how to unencrypt.

**Comment by martingalvan-nordic** [comment]
I agree that this is an extremely useful feature that every proper terminal should have. Leaving a long verbose process running just to come back and see that most of the output was lost is very frustrating.

**Comment by aminomancer** [comment]
Yikes, maybe if extending the limit to Infinity, there could be a profile setting that would allow you to easily wire up a log to file or STDOUT. Unfortunately I need to retrieve the very first line of a script's traceback that's going to be some 400,000 lines long lol. So I will have to write it, though since this is probably a pretty common application, it would be useful for WT to support it with settings.

**Comment by AkikoOrenji** [comment]
for the love of god please fix this. You've provided an amazing terminal (finally comporable to linux Terminator) and then make it next to useless with no external logging or ability to have unlimited history (yes surely 'unlimited' is not that difficult to do in 2022). Off Linux i suppose.....

**Comment by CCCAUCHY** [comment]
It is been long time.
Just make a log file feature please.

**Comment by zadjii-msft** [comment]
Yep. We'll make sure to update this thread when there is an update to share. In the meantime, might I recommend the Subscribe button?
![image](https://user-images.githubusercontent.com/18356694/91237459-5cbb0c80-e700-11ea-9347-b9b1ec2813b1.png)
That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread ☺️

**Comment by zadjii-msft** [comment]
Yep. There are plenty of people following the thread for updates. Just commenting stuff like "are there any updates", "this still hasn't been added", etc, sends everyone who's following the thread a notification, but doesn't _really_ contribute to the discussion of how to actually implement this. I'm pretty liberal with collapsing comments to keep discussions on-topic.

**Comment by lhecker** [comment]
There's no ETA and no updates. If you subscribe to this issue, you'll be notified once that changes.
Work towards this feature is currently ongoing, but it will take a very long time to finally get there.

**Comment by ramapcsx2** [comment]
Hey ho, how about u64 instead of i32? Would already cover most cases :)

**Comment by o-sdn-o** [comment]
You can use [vtm](https://github.com/directvt/vtm) to instantly get this feature inside Windows Terminal.

Run `pwsh` with a scrollback buffer ready to hold, for example, up to 1M lines (will take up ~230 MB of RAM):

```bash
vtm -c "<config><term><scrollback size=1000000/></term></config>" -r vtty pwsh
```

https://github.com/user-attachments/assets/e3791a86-8c06-4232-9108-a5d353f30edc

You can run it in a separate window using `-g`/`--gui` option to see the scrollback buffer state in real time if you need to.

**Comment by effleurager** [comment]
> If there's a way to ping only MS maintainers without affecting all other readers of this, please let me know. Otherwise, I'm afraid that it will keep happening as people lose more time hitting this issue over and over.

You do realise you're the only person affecting this effect?

**Comment by andre-m-coquim-alb** [comment]
Is it possible for an issue as old as this one to be considered a priority and be put out of the backlog, or will it be one of those issues that will remain in backlog forever independently of how many people come here and put a thumbs up or comment on it? Is there anything we, common users can do to give it some kindo of priority?
To me it affects me every day since I have to do compilations that easily pass the current scrollback limit and obligates me to use an alternative terminal, since if an error occurs it will possibly be lost due to the limited scrollback in windows terminal.

**Comment by martingalvan-volue** [comment]
Hi, I just got hit by this issue once again. I saw that @DHowett has set a milestone for this, does that mean there's work being done here?

**Comment by evilZardoz** [comment]
macOS user here transitioning to Windows - this is a significant thing holding me back. The macOS terminal can hold around 2GB of scrollback (before it crashes).  Understand that we need to throw memory at this problem (how efficient would it be? Will I need more than 16 or 32GB RAM in my device for it?).

Glad it's being worked on, at least!

Features for usefulness could include a "save buffer and clear" feature to export the entire buffer to a text file. "Open windows from previous session" would probably want to truncate any buffers that may be stored - macOS terminal does this.

**Comment by lhecker** [comment]
I apologize, I accidentally forgot about finishing the PR, as I was working on another project since earlier this year (msedit). I'll see if I can finish it up in the near term.

**Comment by andre-m-coquim-alb** [comment]
Hi, [lhecker](https://github.com/lhecker), your PR #18290 seems to have failed some checks, can you please take a look when you have time?

**Comment by AldebaraanMKII** [comment]
I just encountered this problem. I could not scroll back to the beginning and had to run my script again to get the initial output. A shame that this is not implemented yet.

**Comment by yuceltoluyag** [comment]
Could this feature be offered as a premium option? I would love to purchase it. @billgates
